### ISSUE TITLE
The 'zerodep' client is running out of connections in connection pool when used heavily?

### ISSUE THREAD

**Issue by kdvolder** [docker-java/docker-java#1466]
Heavy use means multiple threads are making calls concurrently through a single client instacnce.
Some of those calls are streaming data from the logs to display in IDE console and so I guess they are long lived connections.

This is working fine upto a point. Then i it breaks and the client freezes up completely and becomes unable to service any more requests. 

When I inspect what is going on with the debugger I see that it that client calls are 'stuck' in getting a connection from the connection pool.

So my guess  is that it is simply running out of connections in the connection pool. 

Is there a way to increase the size of the pool? 

Appache docs describes how to do this here in section 2.3.3 https://hc.apache.org/httpcomponents-client-ga/tutorial/html/connmgmt.html

However from inspecting the source code of the zerodep client a bit, it looks like the connection manager and its configuration options are not being exposed through Java docker client api. 

It would nice if options to control connection limits could be made configurable. Any chance of this?

**Comment by kdvolder** [comment]
In case it helps. I took a dump of the 'stuck' call in debugger, pasting it below:

```
Thread [Worker-12: AsyncLiveExpression refresh] (Suspended)	
	owns: com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager$1  (id=390)	
	owns: org.springsource.ide.eclipse.commons.livexp.core.ObservableSet$Builder$1  (id=391)	
	waiting for: com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.concurrent.BasicFuture<T>  (id=392)	
	java.lang.Object.wait(long) line: not available [native method]	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.concurrent.BasicFuture<T>.get(long, java.util.concurrent.TimeUnit) line: 101	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager$1.get(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.util.Timeout) line: 278	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.InternalExecRuntime.acquireEndpoint(java.lang.String, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.HttpRoute, java.lang.Object, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.protocol.HttpClientContext) line: 113	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ConnectExec.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain) line: 118	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 51	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 57	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ProtocolExec.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain) line: 165	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 51	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 57	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.HttpRequestRetryExec.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain) line: 93	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 51	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 57	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.RedirectExec.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain) line: 116	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 51	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement$1.proceed(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 57	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ContentCompressionExec.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain) line: 128	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.ExecChainElement.execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.classic.ExecChain$Scope) line: 51	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.InternalHttpClient.doExecute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.HttpHost, com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.protocol.HttpContext) line: 178	
	com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.InternalHttpClient(com.github.dockerjava.zerodep.shaded.org.apache.hc.client5.http.impl.classic.CloseableHttpClient).execute(com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.HttpHost, com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.ClassicHttpRequest, com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.http.protocol.HttpContext) line: 67	
	com.github.dockerjava.zerodep.ZerodepDockerHttpClient(com.github.dockerjava.zerodep.ApacheDockerHttpClientImpl).execute(com.github.dockerjava.transport.DockerHttpClient$Request) line: 149	
	com.github.dockerjava.zerodep.ZerodepDockerHttpClient.execute(com.github.dockerjava.transport.DockerHttpClient$Request) line: 8	
	com.github.dockerjava.core.DefaultInvocationBuilder.execute(com.github.dockerjava.transport.DockerHttpClient$Request) line: 228	
	com.github.dockerjava.core.DefaultInvocationBuilder.get() line: 202	
	com.github.dockerjava.core.DefaultInvocationBuilder.get(com.fasterxml.jackson.core.type.TypeReference<T>) line: 75	
	com.github.dockerjava.core.exec.ListContainersCmdExec.execute(com.github.dockerjava.api.command.ListContainersCmd) line: 44	
	com.github.dockerjava.core.exec.ListContainersCmdExec.execute(com.github.dockerjava.api.command.DockerCmd) line: 15	
	com.github.dockerjava.core.exec.ListContainersCmdExec(com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec<CMD_T,RES_T>).exec(CMD_T) line: 21	
	com.github.dockerjava.core.command.ListContainersCmdImpl(com.github.dockerjava.core.command.AbstrDockerCmd<CMD_T,RES_T>).exec() line: 35	
	org.springframework.ide.eclipse.boot.dash.docker.runtarget.DockerImage.lambda$1(com.github.dockerjava.api.DockerClient) line: 97	
	org.springframework.ide.eclipse.boot.dash.docker.runtarget.DockerImage$$Lambda$1115.1539213400.call() line: not available	
	org.springsource.ide.eclipse.commons.frameworks.core.util.JobUtil.interruptAfter(java.time.Duration, java.util.concurrent.Callable<T>) line: 53	
	org.springframework.ide.eclipse.boot.dash.docker.runtarget.DockerImage.fetchChildren() line: 92	
	<obsolete method in<unknown declaring type>>	
	org.springframework.ide.eclipse.boot.dash.model.remote.GenericRemoteAppElement$$Lambda$958.1245835890.call() line: not available	
	org.springsource.ide.eclipse.commons.livexp.core.ObservableSet$Builder$1.compute() line: 46	
	org.springsource.ide.eclipse.commons.livexp.core.ObservableSet$Builder$1.compute() line: 1	
	org.springsource.ide.eclipse.commons.livexp.core.ObservableSet$Builder$1(org.springsource.ide.eclipse.commons.livexp.core.LiveExpression<V>).refresh() line: 74	
	org.springsource.ide.eclipse.commons.livexp.core.ObservableSet$Builder$1(org.springsource.ide.eclipse.commons.livexp.core.AsyncLiveExpression<T>).syncRefresh() line: 125	
	org.springsource.ide.eclipse.commons.livexp.core.AsyncLiveExpression$1.run(org.eclipse.core.runtime.IProgressMonitor) line: 95	
	org.eclipse.core.internal.jobs.Worker.run() line: 63	
```

**Comment by bsideup** [comment]
Hey @kdvolder,

Make sure that you close the callbacks after using them. It is most probably not the pool that gets exhausted, but the connections that never get returned to the pool

**Comment by kdvolder** [comment]
I beleave they are closed when no longer in use. This was verified by counting and logging and active number of log handlers here:

https://github.com/spring-projects/spring-ide/blob/621121feebc200f3e54c4e3ea5115689075775fc/plugins/org.springframework.ide.eclipse.boot.dash.docker/src/org/springframework/ide/eclipse/boot/dash/docker/runtarget/DockerContainer.java#L472

However, maybe I am misunderstanding something here and doing something wrong. I'm not sure what exactly you mean by 'closing the callback'. 

The callback you refer to, I assume is the `com.github.dockerjava.api.async.ResultCallback<Frame>`. 

That interface has a close method. We implement that interface on our log handler. And its close method is called when we no longer need the connection.

As far as I can tell and have tested, the number of active 'Callbacks' goes back to 0 when it should. And what we are counting here is the number of callbacks which have been created and for which 'close' has not been called. So I think there are no callbacks kept 'open' when they are not in use. 

However what I do not understand is how calling close on the callback is supposed to free up connections in the pool. After all the callback is implemented by us (i.e library user, not the library itself). Our code doesn't control the connection pool and does nothing to free up connections. So thinking this all through, it does seem like something may be missing here. 

Now... I do not know what exactly is wrong though. So can you clarify what you mean by 'closing the callback'. If it just calling the 'close' method on the callback object, then... I think we are doing that already. If it is something else please explain.

BTW: I also have to clarify something about our use of the api.

Even if we did everything correctly, connections may not get returned to the pool for a long time. That is because they are actually being used to stream container output, not because they are not in use. These connections stream data into various IDE console windows, and so that kind of UI means that they must remain active for as long as the user keeps  these windows opened. In practice that typically means streaming connections stay open from the time a container is launched until it is finally stopped. So these connections are naturally rather long lived.

BTW: I found a workaround which is to create dedicated client instances for each 'log streaming' operation. We still use a single shared client for all other 'short lived' request. That seems to 'solve' the issue though it doesn't really feel right to do it this way. And I also fear that it may be degrading performance.

**Comment by kdvolder** [comment]
Okay, I think I made little progress towards understanding proper close of connections.

So the Callback interface also has a 'onStart' method which receives a 'Closeable' argument. So I presume properly closing the connection would mean calling close on that object.

I think we are doing that as well. 

We keep the object here:
https://github.com/spring-projects/spring-ide/blob/621121feebc200f3e54c4e3ea5115689075775fc/plugins/org.springframework.ide.eclipse.boot.dash.docker/src/org/springframework/ide/eclipse/boot/dash/docker/runtarget/DockerContainer.java#L482

And when close on the callback is called, then we close that here:

https://github.com/spring-projects/spring-ide/blob/621121feebc200f3e54c4e3ea5115689075775fc/plugins/org.springframework.ide.eclipse.boot.dash.docker/src/org/springframework/ide/eclipse/boot/dash/docker/runtarget/DockerContainer.java#L459

Is there anything else that perhaps we need to do?

**Comment by kdvolder** [comment]
@bsideup Any thoughts on this you can share? I could be wrong, but as far as I know we are doing everything 'correctly' and the problem is simply that the default pool is not big enough for how we are using it (i.e. our use creates a fair amount of simultaneous long-lived log streaming connections, these connections remain open because they are supposed to continue streaming data).

**Comment by kdvolder** [comment]
ping

**Comment by bsideup** [comment]
Hey @kdvolder,

Thanks for the ping! Sorry, I was unable to find time this week to look into this issue, but I will next week!

**Comment by Arisstath** [comment]
We experience the same issue, we are running out of connections. We launch a lot of concurrent attach sessions.

**Comment by kdvolder** [comment]
@Arisstath Not a great solution, but as mentioned above, a workaround that is working for us is to dedicate a client instance per attach session. I don't think its a great solution, but at least it continues to work, which is better than the client stopping/blocking and failing utterly when too many sessions get opened.

**Comment by bsideup** [comment]
@kdvolder as an alternative workaround, you could pool the clients

**Comment by kdvolder** [comment]
I could but it seems like a bit too complex a thing to implement for what is essentially meant to be just a temporary band-aid until the docker-java client can provide a proper way to control it's internal connection pool size. What would be ideal for our use case is if we could somehow tell the client to increment the pool size by one every time we create a 'long lived' attach session connection (and decrement pool size again after closing such a connection).

**Comment by kdvolder** [comment]
I'm working on PR and have something locally that seems to work for me. It's not yet allowing a 'dynamic' reconfig of the limits. I think this may be a little more tricky to do. However the more simple 'confgure a larger pool up front' approach will probably be good enough.

Here are some things I've learned in the process. 

The limit that we are hitting is a 'per route' limit. I gather all requests to the same docker instance are really seen as being 'the same route'. The default limit for this is set at 5 and this is really rather low. It means that if you open 5 active streaming sessions the client cannot do anything else anymore. So it really isn't too surprising we are hitting those limits easily.

I tested a bit with limits set to 200 and that works fine. Connection pool doesn't really grow much beyond what you actually use (i.e. the limit is a max of 200, but it only creates new connections on demand and they eventually get released again if not used for a while).

Expect a PR sometime soon :-)
